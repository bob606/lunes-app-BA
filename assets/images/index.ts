import AddCircleIcon from './add-circle-icon.svg'
import ArrowLeftCircleIconBlue from './arrow-left-circle-icon-blue.svg'
import ArrowLeftCircleIconWhite from './arrow-left-circle-icon-white.svg'
import ArrowRightCircleIconWhite from './arrow-right-circle-icon-white.svg'
import ArrowRightIcon from './arrow-right-icon.svg'
import BannerGreen from './banner-green.png'
import BannerRed from './banner-red.png'
import BannerYellow from './banner-yellow.png'
import BookIconGrey from './book-icon-grey.svg'
import BookIconWhite from './book-icon-white.svg'
import CheckCircleIconBold from './check-circle-icon-bold.svg'
import CheckCircleIconGreen from './check-circle-icon-green.svg'
import CheckCircleIconWhite from './check-circle-icon-white.svg'
import CheckCircleIcon from './check-circle-icon.svg'
import CheckCloseCircleIconBold from './check-close-circle-icon-bold.svg'
import CheckCloseCircleIcon from './check-close-circle-icon.svg'
import ChevronRight from './chevron-right.svg'
import CloseCircleIconBlue from './close-circle-icon-blue.svg'
import CloseCircleIconBold from './close-circle-icon-bold.svg'
import CloseCircleIconWhite from './close-circle-icon-white.svg'
import CloseCircleIcon from './close-circle-icon.svg'
import CloseIconRed from './close-icon-red.svg'
import CloseIconWhite from './close-icon-white.svg'
import CloseIcon from './close-icon.svg'
import DoubleCheckCircleIconWhite from './double-check-circle-icon-white.svg'
import DoubleCheckIcon from './double-check-icon.svg'
import HeaderCircleIcon from './header-cirlce-icon.svg'
import HeaderLinesIcon from './header-lines-icon.svg'
import HeaderSquareIcon from './header-square-icon.svg'
import HeaderStarIcon from './header-star-icon.svg'
import HeartIconGrey from './heart-icon-grey.svg'
import HeartIconWhite from './heart-icon-white.svg'
import HomeCircleIconBlue from './home-circle-icon-blue.svg'
import HomeCircleIconWhite from './home-circle-icon-white.svg'
import HomeIconGrey from './home-icon-grey.svg'
import HomeIconWhite from './home-icon-white.svg'
import InfoCircleIcon from './info-circle-icon.svg'
import ListIcon from './list-icon.svg'
import LunesIcon from './lunes-icon.svg'
import MenuIcon from './menu-icon.svg'
import OpenLockIcon from './open-lock-icon.svg'
import QRCodeIcon from './qr-code-icon.svg'
import RepeatIcon from './repeat-icon.svg'
import ShareIcon from './share-icon.svg'
import StartIconGrey from './star-icon-grey.svg'
import StartIconWhite from './star-icon-white.svg'
import TrashIcon from './trash-bin-icon.svg'
import TrophyIcon from './trophy-icon.svg'
import VolumeUpCircleIcon from './volume-up-circle-icon.svg'

export {
  AddCircleIcon,
  ArrowLeftCircleIconBlue,
  ArrowLeftCircleIconWhite,
  ArrowRightIcon,
  ArrowRightCircleIconWhite,
  BannerGreen,
  BannerRed,
  BannerYellow,
  BookIconGrey,
  BookIconWhite,
  CheckCircleIcon,
  CheckCircleIconBold,
  CheckCircleIconWhite,
  CheckCircleIconGreen,
  CheckCloseCircleIcon,
  CheckCloseCircleIconBold,
  ChevronRight,
  CloseCircleIcon,
  CloseCircleIconBold,
  CloseCircleIconWhite,
  CloseCircleIconBlue,
  CloseIcon,
  CloseIconWhite,
  CloseIconRed,
  DoubleCheckCircleIconWhite,
  DoubleCheckIcon,
  HeaderCircleIcon,
  HeaderLinesIcon,
  HeaderSquareIcon,
  HeaderStarIcon,
  HeartIconGrey,
  HeartIconWhite,
  HomeCircleIconBlue,
  HomeCircleIconWhite,
  HomeIconGrey,
  HomeIconWhite,
  InfoCircleIcon,
  ListIcon,
  LunesIcon,
  MenuIcon,
  OpenLockIcon,
  QRCodeIcon,
  RepeatIcon,
  ShareIcon,
  StartIconGrey,
  StartIconWhite,
  TrashIcon,
  TrophyIcon,
  VolumeUpCircleIcon
}
