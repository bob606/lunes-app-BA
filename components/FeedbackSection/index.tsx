import {
  React,
  styles,
  View,
  Text,
  IFeedbackProps,
  CorrectIcon,
  IncorrectIcon,
  AlmostCorrectIcon,
} from './imports';

const Feedback = ({
  isCorrect,
  almostCorrect,
  isIncorrect,
  document,
  word,
  article,
}: IFeedbackProps) => (  
  <View style={styles.container}>
    {isCorrect && (
      <View style={[styles.messageContainer, styles.successMessage]}>
        <CorrectIcon />
        <View style={styles.textContainer}>
          <Text style={styles.text}>
            Great, keep it up!{'\n'}
            The Word you filled in is correct.
          </Text>
        </View>
      </View>
    )}

    {isIncorrect && (
      <View style={[styles.messageContainer, styles.failedMessage]}>
        <IncorrectIcon />
        <View style={styles.textContainer}>
          <Text style={styles.text}>
            What a pity! Your entry is incorrect,{'\n'}
            {`the correct answer is: ${document?.article === 'die (Plural)' ? 'die' : document?.article} ${document?.word}`}
          </Text>
        </View>
      </View>
    )}

    {almostCorrect && (
      <View style={[styles.messageContainer, styles.almostCorrectMessage]}>
        <AlmostCorrectIcon />
        <View style={styles.textContainer}>
          <Text style={styles.text}>
            {`Your entry ${article} ${word} is almost correct.\n`}
            Check for upper and lower case.
          </Text>
        </View>
      </View>
    )}
  </View>
);

export default Feedback;
