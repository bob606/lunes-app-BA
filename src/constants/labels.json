{
  "overview": {
    "welcome": "Willkommen bei Lunes!\nLerne Vokabeln für deinen Beruf.",
    "unit": "Lektion",
    "units": "Lektionen",
    "word": "Wort",
    "words": "Wörter",
    "exercises": "Übungen"
  },
  "header": {
    "overview": "Übersicht",
    "overviewExercises": "Übersicht der Übungen",
    "cancelExercise": "Übung beenden"
  },
  "exercises": {
    "overview": {
      "title": "Vokabelübersicht",
      "description": "Alle Wörte anschauen"
    },
    "wordSingleChoice": {
      "title": "Wort auswählen",
      "description": "Wähle das richtige Wort"
    },
    "articleSingleChoice": {
      "title": "Artikel üben",
      "description": "Wähle den richtigen Artikel"
    },
    "write": {
      "title": "Wort schreiben",
      "description": "Wort mit Artikel eingeben",
      "placeholder": {
        "insertAnswer": "Wort mit Artikel eingeben",
        "newTry": "Neuer Versuch"
      },
      "button": {
        "checkInput": "Eingabe überprüfen",
        "showSolution": "Lösung anzeigen",
        "tryLater": "Später versuchen"
      },
      "feedback": {
        "articleMissing": "Ups, deine Eingabe ist nicht vollständig. Füge einen Artikel ein.",
        "wrong": "Die richtige Antwort ist: ",
        "almostCorrect1": "Deine Eingabe ",
        "almostCorrect2": " ist fast richtig.",
        "correct": "Toll, weiter so! \nDeine Eingabe ist richtig."
      }
    },
    "buttons": {
      "next": "Nächstes Wort",
      "showResults": "Ergebnisse anzeigen"
    }
  },
  "results": {
    "resultsOverview": "Ergebnis-Übersicht",
    "feedbackGood": "Toll, weiter so! \nDu hast die Übung sehr gut gemeistert.",
    "feedbackMedium": "Übung macht den Meister, \nversuche es noch einmal!",
    "feedbackBad": "Nicht aufgeben! \nVersuche es noch einmal!",
    "checkEntries": "Eingaben überprüfen",
    "retryExercise": "Übung wiederholen",
    "wrong": "falsche",
    "similar": "ähnliche",
    "correct": "richtige",
    "entries": "Eingaben",
    "viewEntries": "Eingaben anschauen",
    "show": "Zeige",
    "of": "von"
  },
  "cancelModal": {
    "cancelAsk": "Möchtest du diese Übung wirklich beenden?",
    "cancel": "Beenden",
    "continue": "Weiter machen"
  }
}
