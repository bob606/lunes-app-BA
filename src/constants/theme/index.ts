import { COLORS } from './colors'
import { FONTS } from './fonts'
import { SPACINGS } from './spacings'
import { STYLES } from './styles'

export interface Theme {
  colors: typeof COLORS
  fonts: typeof FONTS
  spacings: typeof SPACINGS
  styles: typeof STYLES
}

const theme: Theme = {
  colors: COLORS,
  fonts: FONTS,
  spacings: SPACINGS,
  styles: STYLES,
}

export default theme
