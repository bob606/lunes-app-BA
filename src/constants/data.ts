import { ComponentType } from 'react'
import { SvgProps } from 'react-native-svg'

import {
  AlmostCorrectEntriesIcon,
  CorrectEntriesIcon,
  easy,
  hard,
  IncorrectEntriesIcon,
  mideasy,
  midhard
} from '../../assets/images'
import { RoutesParamsType } from '../navigation/NavigationTypes'
import labels from './labels.json'

export const ExerciseKeys = {
  vocabularyList: 0,
  wordChoiceExercise: 1,
  articleChoiceExercise: 2,
  writeExercise: 3
}
export type ExerciseKeyType = typeof ExerciseKeys[keyof typeof ExerciseKeys]

export interface ExerciseType {
  key: ExerciseKeyType
  title: string
  description: string
  Level: typeof easy
  nextScreen: keyof RoutesParamsType
}

export const EXERCISES: ExerciseType[] = [
  {
    key: ExerciseKeys.vocabularyList,
    title: labels.exercises.vocabularyList.title,
    description: labels.exercises.vocabularyList.description,
    Level: easy,
    nextScreen: 'VocabularyList'
  },
  {
    key: ExerciseKeys.wordChoiceExercise,
    title: labels.exercises.wordChoice.title,
    description: labels.exercises.wordChoice.description,
    Level: mideasy,
    nextScreen: 'WordChoiceExercise'
  },
  {
    key: ExerciseKeys.articleChoiceExercise,
    title: labels.exercises.articleChoice.title,
    description: labels.exercises.articleChoice.description,
    Level: midhard,
    nextScreen: 'ArticleChoiceExercise'
  },
  {
    key: ExerciseKeys.writeExercise,
    title: labels.exercises.write.title,
    description: labels.exercises.write.description,
    Level: hard,
    nextScreen: 'WriteExercise'
  }
]

export const BUTTONS_THEME = {
  light: 'light',
  dark: 'dark',
  noOutline: 'no-outline'
} as const

export type ButtonThemeType = typeof BUTTONS_THEME[keyof typeof BUTTONS_THEME]

interface ArticleType {
  readonly id: number
  readonly value: string
}

export const ARTICLES: ArticleType[] = [
  {
    id: 0,
    value: 'keiner'
  },
  {
    id: 1,
    value: 'Der'
  },
  {
    id: 2,
    value: 'Die'
  },
  {
    id: 3,
    value: 'Das'
  },
  {
    id: 4,
    value: 'Die'
  }
]

export type Article = typeof ARTICLES[number]

export const SIMPLE_RESULTS = { correct: 'correct', incorrect: 'incorrect', similar: 'similar' } as const
export type SimpleResultType = typeof SIMPLE_RESULTS[keyof typeof SIMPLE_RESULTS]

export interface ResultType {
  key: SimpleResultType
  title: string
  Icon: ComponentType<SvgProps>
  order: number
}

export interface Answer {
  word: string
  article: Article
}

export const RESULTS: ResultType[] = [
  {
    key: 'correct',
    Icon: CorrectEntriesIcon,
    title: 'Richtige',
    order: 0
  },
  {
    key: 'similar',
    Icon: AlmostCorrectEntriesIcon,
    title: 'Fast richtige',
    order: 1
  },
  {
    key: 'incorrect',
    Icon: IncorrectEntriesIcon,
    title: 'Falsche',
    order: 2
  }
]

export const numberOfMaxRetries = 3
