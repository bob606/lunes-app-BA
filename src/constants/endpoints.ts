import { Article } from './data'

export interface Discipline {
  id: number
  title: string
  description: string
  icon: string
  numberOfChildren: number
  isLeaf: boolean
  isRoot: boolean
  apiKey?: string
  needsTrainingSetEndpoint: boolean
}

export interface AlternativeWord {
  word: string
  article: Article
}

export interface Image {
  id: number
  image: string
}

export type Images = Image[]

export interface Document {
  id: number
  word: string
  article: Article
  document_image: Images
  audio: string
  alternatives: AlternativeWord[]
}

export type Documents = Document[]

export const ENDPOINTS = {
  disciplines: 'disciplines_by_level',
  disciplinesByGroup: 'disciplines_by_group',
  trainingSet: 'training_set',
  documents: 'documents/:id'
}
