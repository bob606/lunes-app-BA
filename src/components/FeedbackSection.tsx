import React from 'react'
import { View, Text, ImageBackground, StyleSheet } from 'react-native'
import {
  CorrectFeedbackIcon,
  IncorrectFeedbackIcon,
  AlmostCorrectFeedbackIcon,
  incorrect_background,
  hint_background,
  correct_background
} from '../../assets/images'
import { ARTICLES } from '../constants/data'
import { COLORS } from '../constants/colors'
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen'
import { DocumentType } from '../constants/endpoints'

export const styles = StyleSheet.create({
  messageContainer: {
    width: wp('80%'),
    height: hp('9%'),
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: -hp('4%'),
    marginBottom: hp('3%')
  },
  imageBackground: {
    width: wp('80%'),
    height: hp('9%'),
    flexDirection: 'row',
    alignItems: 'center',
    padding: 10
  },
  textContainer: {
    marginHorizontal: 5,
    paddingRight: 15,
    paddingLeft: 5
  },
  text: {
    fontSize: wp('3.5%'),
    fontFamily: 'SourceSansPro-Regular',
    fontWeight: 'normal',
    color: COLORS.lunesBlack
  }
})

export interface FeedbackPropsType {
  secondAttempt: boolean
  result: string
  document?: DocumentType
  input: string
}

const Feedback = ({ result, document, input, secondAttempt }: FeedbackPropsType): JSX.Element | null => {
  const Icon =
    result === 'correct' || result === 'giveUp'
      ? CorrectFeedbackIcon
      : result === 'incorrect' || !secondAttempt
      ? IncorrectFeedbackIcon
      : AlmostCorrectFeedbackIcon

  const background =
    result === 'correct'
      ? correct_background
      : result === 'incorrect' || result === 'giveUp' || !secondAttempt
      ? incorrect_background
      : hint_background

  const message =
    result === 'correct'
      ? 'Toll, weiter so! \nDeine Eingabe ist richtig.'
      : result === 'incorrect' || !secondAttempt
      ? `${result === 'incorrect' ? `Schade, deine Eingabe ist falsch.` : ``} Die richtige Antwort ist: ${
          document?.article?.toLowerCase() === ARTICLES.diePlural ? 'die' : document?.article
        } ${document?.word}`
      : `Deine Eingabe ${input} ist fast richtig. Überprüfe Groß- und Kleinschreibung.`

  return result !== '' || secondAttempt ? (
    <View style={styles.messageContainer}>
      <ImageBackground source={background} style={styles.imageBackground} testID='background-image'>
        <Icon width={28} height={28} />
        <View style={styles.textContainer}>
          <Text numberOfLines={2} ellipsizeMode='tail' style={styles.text}>
            {message}
          </Text>
        </View>
      </ImageBackground>
    </View>
  ) : null
}

export default Feedback
