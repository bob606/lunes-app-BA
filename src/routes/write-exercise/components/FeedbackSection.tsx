import React from 'react'
import {
  AlmostCorrectFeedbackIcon,
  correct_background,
  CorrectFeedbackIcon,
  hint_background,
  incorrect_background,
  IncorrectFeedbackIcon
} from '../../../../assets/images'
import { COLORS } from '../../../constants/colors'
import { heightPercentageToDP as hp, widthPercentageToDP as wp } from 'react-native-responsive-screen'
import { DocumentType } from '../../../constants/endpoints'
import styled from 'styled-components/native'
import labels from '../../../constants/labels.json'

const Background = styled.ImageBackground`
  width: ${wp(80)};
  height: ${hp(9)};
  margin-bottom: 40px;
  flex-direction: row;
  align-items: center;
  padding: 10px;
`

const StyledText = styled.Text`
  width: 100%;
  color: ${COLORS.lunesBlack};
  padding: 0 20px 0 10px;
`

export interface FeedbackPropsType {
  secondAttempt: boolean
  result: string
  document?: DocumentType
  input: string
}

const Feedback = ({ result, document, input, secondAttempt }: FeedbackPropsType): JSX.Element | null => {
  let Icon, background, message
  if (result === 'correct') {
    Icon = CorrectFeedbackIcon
    background = correct_background
    message = labels.exercises.write.feedback.correct
  } else if (result === 'incorrect' || result === 'giveUp' || !secondAttempt) {
    Icon = IncorrectFeedbackIcon
    background = incorrect_background
    message = `${labels.exercises.write.feedback.wrong} „${document?.article.value} ${document?.word}“`
  } else {
    Icon = AlmostCorrectFeedbackIcon
    background = hint_background
    message = `${labels.exercises.write.feedback.almostCorrect1} „${input}“ ${labels.exercises.write.feedback.almostCorrect2}`
  }

  return result !== '' || secondAttempt ? (
    <Background source={background} testID='background-image'>
      <Icon width={28} height={28} />
      <StyledText numberOfLines={2} ellipsizeMode='tail' testID={'feedback-write-exercise'}>
        {message}
      </StyledText>
    </Background>
  ) : null
}

export default Feedback
