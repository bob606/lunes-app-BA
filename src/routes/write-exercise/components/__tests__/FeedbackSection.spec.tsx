/* eslint-disable @typescript-eslint/no-non-null-assertion */
import 'react-native'
import React from 'react'
import FeedbackSection, { FeedbackPropsType } from '../FeedbackSection'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { ARTICLES } from '../../../../constants/data'
import { AlmostCorrectFeedbackIcon, CorrectFeedbackIcon, IncorrectFeedbackIcon } from '../../../../../assets/images'
import labels from '../../../../constants/labels.json'

describe('Feedback section', () => {
  const defaultFeedbackProps: FeedbackPropsType = {
    input: '',
    result: '',
    secondAttempt: false,
    document: {
      alternatives: [],
      article: ARTICLES[0],
      audio: '',
      id: 0,
      document_image: [],
      word: ''
    }
  }

  it('should render without issues', () => {
    const component = shallow(<FeedbackSection {...defaultFeedbackProps} />)
    expect(toJson(component)).toMatchSnapshot()
  })

  it('should not render when there is no result and secondAttempt is false', () => {
    const feedbackProps: FeedbackPropsType = {
      ...defaultFeedbackProps
    }

    const component = shallow(<FeedbackSection {...feedbackProps} />)
    expect(component.isEmptyRender()).toBe(true)
  })

  it('should render success feedback when the result is correct', () => {
    const feedbackProps: FeedbackPropsType = {
      ...defaultFeedbackProps,
      result: 'correct'
    }
    const message = labels.exercises.write.feedback.correct
    const icon = <CorrectFeedbackIcon width={28} height={28} />

    const component = shallow(<FeedbackSection {...feedbackProps} />)
    expect(component.props().children[1].props.children).toBe(message)
    expect(component.contains(icon)).toBe(true)
  })

  it('should render error feedback when the result is incorrect and the second attempt is false', () => {
    const feedbackProps: FeedbackPropsType = {
      ...defaultFeedbackProps,
      result: 'incorrect',
      document: {
        article: ARTICLES[1],
        word: 'Winkel',
        audio: '',
        alternatives: [],
        id: 1,
        document_image: []
      }
    }
    const message = `${labels.exercises.write.feedback.wrong} „${feedbackProps.document!.article.value} ${
      feedbackProps.document!.word
    }“`
    const icon = <IncorrectFeedbackIcon width={28} height={28} />

    const component = shallow(<FeedbackSection {...feedbackProps} />)
    expect(component.props().children[1].props.children).toBe(message)
    expect(component.contains(icon)).toBe(true)
  })

  it('should render almost correct feedback when the result is similar', () => {
    const feedbackProps: FeedbackPropsType = {
      ...defaultFeedbackProps,
      result: 'similar',
      input: 'der winkel',
      secondAttempt: true
    }
    const message = `${labels.exercises.write.feedback.almostCorrect1} „${feedbackProps.input}“ ${labels.exercises.write.feedback.almostCorrect2}`
    const icon = <AlmostCorrectFeedbackIcon width={28} height={28} />

    const component = shallow(<FeedbackSection {...feedbackProps} />)
    expect(component.props().children[1].props.children).toBe(message)
    expect(component.contains(icon)).toBe(true)
  })
})
