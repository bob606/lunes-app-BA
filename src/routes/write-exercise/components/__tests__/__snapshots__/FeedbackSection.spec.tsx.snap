// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback section should render without issues 1`] = `""`;
